### PR TITLE
strict internal libc compat

### PR DESCRIPTION
**Pull request by fbrausse**
This PR enforces usage of headers of ESBMC's internal libc for `c2goto` as well as `esbmc` when parsing source files (with the clang-c-frontend). If this approach turns out to be successful (no clashes with system headers on any supported OS),
- we can remove unchanged copies of headers such as bits/pthreadtypes.h, see #600, and
- we will get compiler errors for incompatibilities between the OS headers and ESBMC's; thereby
- discrepancies between any pre-built operational models and the translated user sources should be minimized, that is, for any library we (want to) model (e.g., POSIX threads), the sources will be translated with types corresponding to ESBMC's libc.

~It is work-in-progress at the moment.~

### CODE REVIEW

**Comment by fbrausse** [comment]
Linux:
>  The following tests FAILED:
-	263 - regression-esbmc-unix2-11_olstate (Failed)
  Broken test-case in my opinion: can't #define uint8_t and uint16_t to int and expect standard headers to agree.
-	289 - regression-esbmc-unix2-23_picosat-846_01 (Failed)
	290 - regression-esbmc-unix2-23_picosat-846_02 (Failed)
	292 - regression-esbmc-unix2-23_picosat-846_04 (Failed)
	293 - regression-esbmc-unix2-23_picosat-846_05 (Failed)
  need to fix `union pthread_attr_t` in this PR.
-	1552 - regression-esbmc-cpp/cpp-cpp_with_pthread_01 (Failed)
	1553 - regression-esbmc-cpp/cpp-cpp_with_pthread_02 (Failed)
  `KNOWNBUG` -> `CORE`

*Edit*: done.

**Comment by fbrausse** [comment]
Obviously I'll clean up the history to remove my failed attempts. Many of those have been to figure out what is really meant by setting include paths (e.g. for the macOS build). The following convention is introduced in the clang-c-frontend instead of just using `-I`:
- ESBMC's libc path is the first `-isystem` parameter to clang
- Clang's header path is the second `-isystem` parameter to clang
- the macOS build now uses `c2goto --sysroot ${C2GOTO_INCLUDE_DIR}/../..` -- this is obviously in need of cleanup, ie, rename the CMake parameter; note that `--sysroot` replaces the directories in step 5 below

We move to `-isystem` instead of `-I` in order to bring ESBMC's behaviour regarding include-paths closer to the order in which clang/gcc handle #include-s, thereby freeing `-I` for users alone:
1. `"..."` is looked up in the directory of the source
2. `"..."` is looked up in the `-iquote` list from left to right
3. both forms `"..."` and `<...>` are looked up in the `-I` list from left to right
4. both forms are looked up in the `-isystem` list from left to right
5. look up in standard system directories
6. look up in the `-idirafter` list from left to right

**Comment by fbrausse** [comment]
PS: I should add that accepting this PR is a step towards supporting the different OS libc-s more whole-heartedly - or replacing them by an own one, see also #183.

**Comment by lucasccordeiro** [comment]
@fbrausse: Is this PR ready to review?

**Comment by fbrausse** [comment]
Yes.

**Review by lucasccordeiro — APPROVED** [review]
LGTM.

**Comment by lucasccordeiro** [comment]
@fbrausse: The following test cases are failing in MacOS. Is it a bug or a feature?

The following tests FAILED:
	 27 - regression-esbmc-unix-00_account_01 (Failed)
	 28 - regression-esbmc-unix-00_account_02 (Failed)
	 34 - regression-esbmc-unix-00_atomicity06 (Failed)
	 36 - regression-esbmc-unix-00_bbuf_01 (Failed)
	 37 - regression-esbmc-unix-00_bbuf_02 (Failed)
	 38 - regression-esbmc-unix-00_bbuf_02_new (Failed)
	 45 - regression-esbmc-unix-00_race07 (Failed)
	 52 - regression-esbmc-unix-00_race14 (Failed)
	 55 - regression-esbmc-unix-00_race17 (Failed)
	 56 - regression-esbmc-unix-00_race18 (Failed)
	 57 - regression-esbmc-unix-00_race19 (Failed)
	 58 - regression-esbmc-unix-00_race20 (Failed)
	 73 - regression-esbmc-unix-01_malloc_02 (Failed)
	 76 - regression-esbmc-unix-01_malloc_05 (Failed)
	 77 - regression-esbmc-unix-01_malloc_06 (Failed)
	 81 - regression-esbmc-unix-01_malloc_10 (Failed)
	 87 - regression-esbmc-unix-01_malloc_16 (Failed)
	125 - regression-esbmc-unix-03_boundedBuffer (Failed)
	126 - regression-esbmc-unix-03_bzip2smp_comb_01 (Failed)
	127 - regression-esbmc-unix-03_bzip2smp_comb_01_new (Failed)
	130 - regression-esbmc-unix-03_falcon (Failed)
	140 - regression-esbmc-unix-03_twostage_01 (Failed)
	141 - regression-esbmc-unix-03_twostage_02 (Failed)
	145 - regression-esbmc-unix-03_wronglock_01 (Failed)
	146 - regression-esbmc-unix-03_wronglock_02 (Failed)
	147 - regression-esbmc-unix-03_wronglock_03 (Failed)
	149 - regression-esbmc-unix-04_bzip2smp_03 (Failed)
	165 - regression-esbmc-unix-github_397_3 (Failed)
	166 - regression-esbmc-unix-github_397_4 (Failed)
	176 - regression-esbmc-unix2-01_pthread11 (Failed)
	177 - regression-esbmc-unix2-01_pthread12 (Failed)
	205 - regression-esbmc-unix2-01_pthread38 (Failed)
	216 - regression-esbmc-unix2-01_pthread48 (Failed)
	219 - regression-esbmc-unix2-01_pthread50 (Failed)
	220 - regression-esbmc-unix2-01_pthread51 (Failed)
	221 - regression-esbmc-unix2-01_pthread52 (Failed)
	222 - regression-esbmc-unix2-01_pthread53 (Failed)
	223 - regression-esbmc-unix2-01_pthread54 (Failed)
	234 - regression-esbmc-unix2-01_pthread64 (Failed)
	235 - regression-esbmc-unix2-01_pthread65 (Failed)
	237 - regression-esbmc-unix2-01_pthread67 (Failed)
	260 - regression-esbmc-unix2-11_lu-fig2.fixed (Failed)
	305 - regression-esbmc-unix2-github_213 (Failed)
Errors while running CTest
Error: Process completed with exit code 8.

**Comment by fbrausse** [comment]
> The following test cases are failing in MacOS. Is it a bug or a feature?

The logs show pretty much the discrepencies between ESBMC's headers and (in this case, since our headers come from glibc) the OS X headers. `union pthread_attr_t` vs. `struct _opaque_pthread_attr_t`, etc. These show explicitely where assumptions before had silently worked - just because the space occupied in memory has not been overrun. Assumptions were:
- internal operational models are defined based on the glibc-provided pthread types defined in `c2goto/headers`.
- user code (regression tests in this case) have been translated using system definitions - which differ between libc-s.

I don't expect this PR to be merged quickly, I suspect it is going to take time to sort this out - in particular since I don't have an OS X machine available.

Essentially, I see 2 ways to fix these issues:
1. get rid of most libc-specific headers (like `bits/pthreadtypes.h`, but not compiler-provided ones like `stddef.h`) and model the behaviour for every supported libc/OS separately (e.g., via `#ifdef __APPLE__`, `#ifdef __GLIBC__`, etc. *edit* Free-/Cheribsd to come)
2. settle for one specific implementation (does not have to be an existing one, just one that follows all assumptions made by standards like POSIX) and provide headers defining the concrete types used in the (single) operational model.

**Comment by fbrausse** [comment]
The choice between 1. and 2. (which I as a newcomer cannot make) affects whether ESBMC is to support code that (preprocessor-)branches on different libc-s or we only want to support code that is oblivious to the concrete library implementation (i.e., it doesn't care how pthread_attr_t is defined, as long as the API works) and thus accepts ESBMC's models independent of the system.

Obviously when there is no standard - like e.g, OpenSSL 1.1.x vs. 3.y vs. LibreSSL vs..., choice 1. is better for users while putting a burden on ESBMC's maintainers. On the other hand standards exist for a reason - and following a moving target like the *SSL implementations could be considered a user's choice when employing a general C/C++ model checker.

**Review by kunjsong01 — APPROVED** [review]

File: src/clang-c-frontend/clang_c_language.cpp [hunk 1]
```diff
@@ -43,7 +43,15 @@ void clang_c_languaget::build_compiler_args(const std::string &tmp_dir)
...
 
-  compiler_args.push_back("-I" + tmp_dir);
+  const std::string *libc_headers = internal_libc_header_dir();
+  if(libc_headers)
+  {
+    compiler_args.push_back("-isystem");
+    compiler_args.push_back(*libc_headers);
+  }
+
+  compiler_args.push_back("-isystem");
```
Comment (kunjsong01) [inline src/clang-c-frontend/clang_c_language.cpp:53]: Same as line 49. pehaps make it common?

**Comment by fbrausse** [comment]
@lucasccordeiro explained to me in private conversation:
> the operational models are an abstract representation of the original libraries
So, as long as we capture the original semantics of the library, we should be okay

which I interpret as going for ~option 1~ *edit:* sorry, misread: option 2 above. Incidentally, the regressions failing on CI for #615 profit from this as well.

File: src/clang-c-frontend/clang_c_language.cpp (hunk 1 above)
Comment (fbrausse) [reply src/clang-c-frontend/clang_c_language.cpp:53]: We would have the same `if` plus another std::vector. I don't think it's worth it at this point. We don't even expose `-isystem` as a cmdline parameter at this point.

**Comment by fbrausse** [comment]
The macOS issues are fixed now by simplifying its types, and removing all architecture-specific #ifdefs.

**Review by lucasccordeiro — APPROVED** [review]
@fbrausse: This is a very nice piece of work! Well done!

I believe this PR will improve the performance of our concurrency checking quite a lot!

Can I ask you to run this PR over the concurrency category only? 

You can use these parameters:

1) 
  if concurrency:
    command_line += " --no-por --context-bound 2  "

2)
  if concurrency:
    command_line += " --no-por --context-bound 2  --no-goto-merge "

This PR contains the latest results for comparison: https://github.com/esbmc/esbmc/pull/617

**Comment by lucasccordeiro** [comment]
@fbrausse: could you also please rebase this branch?

**Comment by lucasccordeiro** [comment]
This PR with  `--incremental-bmc --no-por --context-bound 2 --no-goto-merge`:

```
Statistics:            763 Files
  correct:             336
    correct true:       69
    correct false:     267
  incorrect:             0
    incorrect true:      0
    incorrect false:     0
  unknown:             427
  Score:               405 (max: 1160)
```

Master with `--incremental-bmc --no-por --context-bound 2 --no-goto-merge`:

```
Statistics:            763 Files
  correct:             336
    correct true:       69
    correct false:     267
  incorrect:             0
    incorrect true:      0
    incorrect false:     0
  unknown:             427
  Score:               405 (max: 1160)
```

**Comment by lucasccordeiro** [comment]
This PR with  `--incremental-bmc --no-por --context-bound 2 --no-goto-merge`:

```
Statistics:            763 Files
  correct:             336
    correct true:       69
    correct false:     267
  incorrect:             0
    incorrect true:      0
    incorrect false:     0
  unknown:             427
  Score:               405 (max: 1160)
```

Master with `--incremental-bmc --no-por --context-bound 2 --no-goto-merge`:

```
Statistics:            763 Files
  correct:             336
    correct true:       69
    correct false:     267
  incorrect:             0
    incorrect true:      0
    incorrect false:     0
  unknown:             427
  Score:               405 (max: 1160)
```

**Pull request merged**